### DISCUSSION TITLE
Can I get a little clarification over my little doubts about GGUF? #732

### DISCUSSION THREAD

**Discussion by AayushSameerShah** [ggml-org/llama.cpp#3246] [Q&A]
Hello, community!
Recently I have witnessed the rise of `Llama.cpp` and how it has managed to let anyone use LLM on their personal computer. I am having some **basic gaps in knowledge** to wrap my mind around this surge.

# 1️⃣ Running on CPU
I am willing to use the `llama-2-chat` model and in quantized format. And only thing is my CPU. It is true that it can run on CPU, but is the speed drastically slowed down? 

And specially **is there anything to accelerate the speed** while running the 4bit quantized model on CPU?

# 2️⃣ What is the use of BLAS and all other jargons?
I mean, when I went through installing llama.cpp, it had many, I mean many steps to go by. BLAS looked like it could give acceleration in the inference.
What is `mpirun`? I mean there are a lot of things...

The question is: **Will BLAS still work whole using CPU?** Is it required?
# 3️⃣  How is this different than Ctransformers?
Okay, there is `llama.cpp` and there are other implementations in python: `llama-cpp-python`, java: `java-llama.cpp`... but what is this CTransformers? How is it different than the llama.cpp?

# 4️⃣ Can I get a filtered, step-by-step guide for installation on Windows?
The README of llama.cpp is pretty clear, but it has all ways scattered for all OS in the single page and it becomes hard to navigate for your purpose.

My purpose:
- Run **LLama-2-chat** model (https://huggingface.co/TheBloke/Llama-2-7B-Chat-GGML)
- On **CPU** only
- On **Windows**
- With Python

Now, what things should I install to get the maximum inference speed? Will you please guide me through that?

___

I know I am asking a lot, but If you can provide me a simple and straight guide to get the maximum speed for my requirements, it will be amazing!

Apologies for the noobie questions,
Thanks! 🙏🏻

**Comment by ianscrivener** [comment]
Yes, CPU obly will be much slower.
Suggest you try it and see if the performance is acceptable. It wil really depend on your CPU, RAM and expectations.

As I understand, some CPUs support BLAS. 
There is also OpenBLAS - https://github.com/xianyi/OpenBLAS/blob/develop/README.md

This is all very new tech... not much documentation there yet.

Good luck!

**Reply by AayushSameerShah** [reply]
Thanks for the response mate! 
I am a little unsure about the differences between llama.cpp and ctransformer libraries. When to use what? I mean, llamma cpp has many wrappers availalbe in different languages, such as `llama-cpp-python` in python, but... what is ctransformers then?

Please enlighten me.
Thanks.

**Reply by staviq** [reply]
An LLM model is a "data dump", a file, it doesn't do anything by itself. That model file, basically contains a giant mathematical formula, which takes numbers in and spits numbers out. You need two things to use a model: something that can turn your input text into numbers, because that is the only thing a model understands, and that is called tokenizer/detokenizer, and something that can calculate that mathematical formula with the numbers from tokenizer, and give it back to the detokenizer so you get the result in a text form.

Llama.cpp is a program, which can turn your text into numbers, pass it through that mathematical formula ( the model ) and turn the resulting numbers back to text. The core part of llama.cpp is about calculating that mathematical formula, as fast as possible, either on the CPU, or the GPU.

Transformer libraries, and similar other projects, are mostly just different projects doing the same time thing, but typically each project has it's pros and cons. Most other projects require a GPU, while llama.cpp doesn't, it can run on pretty much anything.

Bindings for llama.cpp are simply interfaces between llama.cpp and other programming languages, so you can write your own code or scripts which can use an LLM in the background.

**Reply by AayushSameerShah** [reply]
Yeah, thanks for an amazing response.
I am little unsure about using BLAS and other things I found on the README page. Since I am using only CPU, will installing BLAS help me?

I saw the text:
```
The above will distribute the computation across 2 processes on the first host and 1 process on the second host. Each process will use roughly an equal amount of RAM. Try to keep these numbers small, as inter-process (intra-host) communication is expensive.

Finally, you're ready to run a computation using mpirun:
```
Which explicitly tells to use somethings else to distribute the computation across multiple processors. That logically means, the library by default doesn't distribute the processes.

Then it goes over `BLAS`, `BLIS`, `Intel MKL` and so on. It isn't clear *(to me)* which one should I be using and how each it different. *(of course one should do more research to get the understanding on these BLAS and BLIS thing but you know...)*

That's what confused me... 😕

**Reply by staviq** [reply]
Here's a quite descriptive article on BLAS and different variants, basically those are helper libraries for math functions: https://en.wikipedia.org/wiki/Basic_Linear_Algebra_Subprograms#Implementations

MPI is a way ( a "framework" ) to run a program on multiple computers simultaneously, sort of "merging" multiple computers into one virtual CPU, but that's a very big simplification.
To use it, you don't run llama.cpp, you use `mpirun` program with a parameter telling it to run llama.cpp, which connects to other ( predefined ) computers, and starts several copies of llama.cpp, and synchronizes them so they work "on the same thing".

MPI is not a magic solution to make things faster unfortunately, in many cases, especially when using small models, MPI will actually be slower than just running llama.cpp normally.

Realistically, to use MPI effectively, you would need something like 10Gbit network connection between hosts, because with it, whatever was being transferred between the CPU/GPU/RAM, now needs to be transferred over network connection.

**Comment by ianscrivener** [comment]
llama.cpp is a performant "LLM inference engine" - it does the hard work.

ctransformers is a python binding for llama.cpp - allows you to use llama.cpp in your python code/apps. There's also [llama-cpp-python](https://github.com/abetlen/llama-cpp-python). I use llama-cpp-python. Both ctransformers and llama-cpp-python can be used by UI apps and cool LLM apps like langchain and llamaindex.

There are also Rust, Go and node.js bindings for llama.cpp.

**Reply by AayushSameerShah** [reply]
Alright, so I can think ctransformers as an alternative to the llama-cpp-python.
Thanks 😄

**Reply by ianscrivener** [reply]
Yup

**Reply by ianscrivener** [reply]
**Correction:** it seems that ctransformers uses GGML/GGUF... but **does not** actually use llama.cpp inside

It looks (kinda of) like of a combination of llama.cpp + llama-cpp-python

Like I said - I use llama.cpp + llama-cpp-python... not very familiar with Ctransformers

Guessing llama.cpp would be more performant that ctransformers...

**Reply by AayushSameerShah** [reply]
Yeah, thanks. I have made up my mind around using [llama-cpp-python](https://github.com/abetlen/llama-cpp-python) already, and as you've said it looks performant and so on.

So here's what *(I think)* I am gonna do:
- pip install llama-cpp-python

And haha, that's it. I don't think I should install any BLAS or anything. I mean if I need then please tell me, **because** in the documentation I can only see 2 references to BLAS while using the [llama-cpp-python API doc](https://abetlen.github.io/llama-cpp-python/#llama_cpp.llama_cpp.GGML_USE_CUBLAS):
1. `LLAMA_MAX_DEVICES = GGML_CUDA_MAX_DEVICES if GGML_USE_CUBLAS else ctypes.c_int(1)`
2. `GGML_USE_CUBLAS = hasattr(_lib, 'ggml_init_cublas')`

And I am not sure if there are any other things to install *(use)* to accelerate performance than BLAS - also don't know how to install it with llama-cpp-python as said only 2 mentions above.

I know I sound too confusing, but trying to show the confusion that I have currently.
Anyways, thank you mate for your prompt responses. It will be very helpful to me 🙏🏻

**Reply by ianscrivener** [reply]
also Andrei Abetlen from [llama-cpp-python](https://github.com/abetlen/llama-cpp-python) does a good job of keeping current with llama.cpp ... so llama.cpp improvements are available promptly in llamap-python-cpp

**Reply by ianscrivener** [reply]
I am on MacOS... but I'd check your CPU for BLAS support, check the BLAS & OpenBLAS doco and choose the best compatible option. Be sure to test building (make) with both to test relative performance.
